digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/NodesAndJunctionsService.scala

```diff
@@ -18,12 +19,70 @@ class NodesAndJunctionsService(roadwayDAO: RoadwayDAO, roadwayPointDAO: RoadwayP
 
   def withDynTransaction[T](f: => T): T = OracleDatabase.withDynTransaction(f)
 
+  def withDynTransactionNewOrExisting[T](f: => T): T = OracleDatabase.withDynTransactionNewOrExisting(f)
+
   def withDynSession[T](f: => T): T = OracleDatabase.withDynSession(f)
 
   private val logger = LoggerFactory.getLogger(getClass)
 
   val roadwayAddressMapper = new RoadwayAddressMapper(roadwayDAO, linearLocationDAO)
 
+  def update(node: Node, junctionsIds: Seq[Long], nodePointIds: Seq[Long], username: String = "-"): Option[String] = {
+    withDynTransaction {
+      try {
+        addOrUpdateNode(node, username) match {
+          case Some(err) => return Some(err)
+          case _ => None
+        }
+        detachJunctionsFromNode(junctionsIds, username)
+        detachNodePointsFromNode(nodePointIds, username)
+      } catch {
+        case e: Exception => Some(e.getMessage)
+      }
+    }
+  }
+
+  def addOrUpdateNode(node: Node, username: String = "-"): Option[String] = {
+    withDynTransactionNewOrExisting {
+      try {
+        if (node.id == NewIdValue) {
+          nodeDAO.create(Seq(node), username)
+        } else {
+          val old = nodeDAO.fetchById(node.id)
+          if (old.isDefined) {
+            val oldStartDate = old.get.startDate.withTimeAtStartOfDay
+            val newStartDate = node.startDate.withTimeAtStartOfDay
+            if (node.name != old.get.name || old.get.nodeType != node.nodeType || oldStartDate != newStartDate
+              || old.get.coordinates != node.coordinates) {
+
+              // Update the node information
+              if (old.get.nodeType != node.nodeType && oldStartDate != newStartDate) {
+
+                // Check that new start date is not earlier than before
+                if (newStartDate.getMillis < oldStartDate.getMillis) {
+                  return Some("Solmun uusi alkupäivämäärä ei saa olla ennen nykyistä alkupäivämäärää.")
+                }
+
+                // Create a new history layer when the node type has changed
+                nodeDAO.create(Seq(old.get.copy(id = NewIdValue, endDate = Some(node.startDate.minusDays(1)))), username)
+                nodeDAO.create(Seq(node.copy(id = NewIdValue)), username)
+
+              } else {
+                nodeDAO.create(Seq(node.copy(id = NewIdValue)), username)
+              }
+              nodeDAO.expireById(Seq(node.id))
+            }
+          } else {
+            return Some("Päivitettävää solmua ei löytynyt")
```

Comment:
should we add the message to package.scala?

Reply:
- val **NodeNotFoundErrorMessage** = "Päivitettävää solmua ei löytynyt."

- val **NodeStartDateUpdateErrorMessage** = "Solmun uusi alkupäivämäärä ei saa olla ennen nykyistä alkupäivämäärää."